corehq/apps/sso/tests/test_backends.py

```diff
@@ -0,0 +1,225 @@
+from django.contrib import auth
+from django.test import TestCase, RequestFactory
+
+from corehq.apps.domain.models import Domain
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+from corehq.apps.sso.tests import generator
+from corehq.apps.users.models import WebUser
+
+
+class TestSsoBackend(TestCase):
+    """
+    This ensures that authentication through the SsoBackend only succeeds
+    when a samlSessionIndex is present and the Identity Provider, its
+    associated email domains, and the username all meet required criteria.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.account = generator.get_billing_account_for_idp()
+        cls.domain = Domain(
+            name="vaultwax-001",
+            is_active=True
+        )
+        cls.domain.save()
+
+        # this will be the user that's "logging in" with SAML2 via the SsoBackend
+        cls.user = WebUser.create(
+            cls.domain.name, 'b@vaultwax.com', 'testpwd', None, None
+        )
+        cls.idp = generator.create_idp('vaultwax', cls.account)
+        cls.idp_two = generator.create_idp('vwx-link', cls.account)
+
+    @classmethod
+    def tearDownClass(cls):
+        IdentityProvider.objects.all().delete()
+        cls.user.delete(None)
+        cls.domain.delete()
+        cls.account.delete()
+        super().tearDownClass()
+
+    def setUp(self):
+        super().setUp()
+        # Identity Provider is_active is always True by default for each test
+        self.idp.is_active = True
+        self.idp.save()
+
+        self.request = RequestFactory().get('/sso/test')
+        self.request.session = {}
+
+    def tearDown(self):
+        AuthenticatedEmailDomain.objects.all().delete()
+        super().tearDown()
+
+    def _set_saml_session_index(self):
+        self.request.session['samlSessionIndex'] = '_7c84c96e-8774-4e64-893c-06f91d285100'
+
+    def test_backend_failure_without_idp_slug(self):
+        """
+        SsoBackend should fail to move past the first check because an idp_slug
+         was not passed to authenticate()
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username
+        )
+        self.assertIsNone(user)
+        with self.assertRaises(AttributeError):
```

Comment:
Curious about the semantics here. If I don't specify an identity provider slug, why don't I receive an error message like the other failures?

Reply:
I mentioned this in another comment, but it's because that backend was not appropriate for the given set of `kwargs`. django will try each authentication backend in order when `authenticate` is called. If a slug wasn't present, it means `authenticate` was called somewhere else and it was unrelated to sso.